chainer/function.py

```diff
@@ -174,6 +177,43 @@ def __call__(self, *inputs):
             return ret[0]
         return ret
 
+    def _check_data_type_forward(self, in_data):
+        in_type = type_check.get_types(in_data, 'in_types', False)
+        self.check_type_forward(in_type)
+
+    def _check_data_type_backward(self, in_data, grad_data):
+        in_type = type_check.get_types(in_data, 'in_types', False)
+        grad_type = type_check.get_types(grad_data, 'grad_types', True)
+        self.check_type_backward(in_type, grad_type)
+
+    def check_type_forward(self, in_types):
+        """Checks types of input data before forward propagation.
+
+        Before :meth:`forward` is called, this function is called.
+        You need to validate types of input data in this function
+        using type-check utilities.
+
+        Args:
+            in_types: :class:`TypeInfoTuple` object which contains type
+                information of input data for :meth:`forward`.
+        """
+        pass
+
+    def check_type_backward(self, in_types, grad_types):
+        """Checks types of gradient data before back propagation.
```

Comment:
We need comments somewhere that `backward` should be called after `forward` is called